OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs

```diff
@@ -828,6 +828,53 @@ internal static void UpgradeWeaponRules(ModData modData, int engineVersion, ref
 					}
 				}
 
+				// Valid-/InvalidImpactTypes were removed from CreateEffectWarhead, which uses Valid-/InvalidTargets instead now
+				if (engineVersion < 20170625)
+				{
+					if (node.Key.StartsWith("Warhead", StringComparison.Ordinal) && node.Value.Value == "CreateEffect")
+					{
+						var validImpactTypes = node.Value.Nodes.FirstOrDefault(n => n.Key == "ValidImpactTypes");
+						var invalidImpactTypes = node.Value.Nodes.FirstOrDefault(n => n.Key == "InvalidImpactTypes");
+						if (validImpactTypes != null)
+						{
+							Console.WriteLine("CreateEffectWarhead now uses Valid-/InvalidTargets instead of Valid-/InvalidImpactTypes.");
+							Console.WriteLine("Please check whether you need to make manual adjustments.");
+
+							var validTargets = "";
```

Comment:
It would be cleaner to make this a `List<string>` and then use `FieldSaver.FormatValue` to turn it into a yaml string at the end.